src/backend/distributed/worker/worker_create_if_not_exist.c

```diff
@@ -0,0 +1,210 @@
+
+
+static bool
+object_from_create_exists(Node *parseTree)
+{
+	switch (parseTree->type)
+	{
+		case T_CompositeTypeStmt:
+		{
+			return CompositeTypeExists(castNode(CompositeTypeStmt, parseTree));
+		}
+
+		case T_CreateEnumStmt:
+		{
+			return EnumTypeExists(castNode(CreateEnumStmt, parseTree));
+		}
+
+		default:
+		{
+			/*
+			 * should not be reached, indicates the coordinator is sending unsupported
+			 * statements
+			 */
+			ereport(ERROR, (errmsg("unsupported statement to check existence for"),
+							errhint("The coordinator send an unsupported command to the "
+									"worker")));
+			return false;
+		}
+	}
+}
+
+
+static DropStmt *
+drop_stmt_from_object_create(Node *createStmt)
+{
+	switch (nodeTag(createStmt))
+	{
+		case T_CompositeTypeStmt:
+		{
+			return CompositeTypeStmtToDrop(castNode(CompositeTypeStmt, createStmt));
+		}
+
+		case T_CreateEnumStmt:
+		{
+			return CreateEnumStmtToDrop(castNode(CreateEnumStmt, createStmt));
+		}
+
+		default:
+		{
+			/*
+			 * should not be reached, indicates the coordinator is sending unsupported
+			 * statements
+			 */
+			ereport(ERROR, (errmsg("unsupported statement to check existence for"),
+							errhint("The coordinator send an unsupported command to the "
+									"worker")));
+		}
+	}
+}
+
+
+/*
+ * worker_create_if_not_exists(sqlStatement text)
+ *
+ * The sqlStatement is parsed and interpreted to find the relation it is supposed to
+ * create. The command will only be executed if the relation does not already exist. The
+ * relation is not checked to see if they are different. It is assumed that if the
+ * relation exists it should not be created.
+ */
+Datum
+worker_create_if_not_exists(PG_FUNCTION_ARGS)
+{
+	text *sqlStatementText = PG_GETARG_TEXT_P(0);
+	const char *sqlStatement = text_to_cstring(sqlStatementText);
+
+	Node *parseTree = ParseTreeNode(sqlStatement);
+
+	if (object_from_create_exists(parseTree))
+	{
+		PG_RETURN_BOOL(false);
+	}
+
+	CitusProcessUtility(parseTree, sqlStatement, PROCESS_UTILITY_TOPLEVEL, NULL,
+						None_Receiver, NULL);
+
+	/* type has been created */
+	PG_RETURN_BOOL(true);
+}
+
+
+Datum
+worker_create_or_replace(PG_FUNCTION_ARGS)
+{
+	text *sqlStatementText = PG_GETARG_TEXT_P(0);
+	const char *sqlStatement = text_to_cstring(sqlStatementText);
+
+	Node *parseTree = ParseTreeNode(sqlStatement);
+
+	/*
+	 * since going to the drop statement might require some resolving we will do a check
+	 * if the type actually exists instead of adding the IF EXISTS keyword to the
+	 * statement.
+	 */
+	if (object_from_create_exists(parseTree))
+	{
+		DropStmt *dropStmtParseTree = NULL;
+
+		/*
+		 * there might be dependencies left on the worker on this type, these are not
+		 * managed by citus anyway so it should be ok to drop, thus we cascade to any such
+		 * dependencies
+		 */
+		dropStmtParseTree = drop_stmt_from_object_create(parseTree);
+		dropStmtParseTree->behavior = DROP_CASCADE;
```

Comment:
Would not this fail if `dropStmtParseTree` is `NULL`? Should we move this line after the NULL-check below?

Reply:
makes sense

this function however will change significantly, where it will probably not drop anymore, but instead rename it (to prevent accidental data destruction) if the type is not already the same as the type you are creating.